### PR TITLE
Liquify

### PR DESCRIPTION
**Pull request by TurboGit**
This is the first pull-request for the liquify iop. I'm not the original author but I've taken over the clean-up to have this integrated into darktable. At this stage it looks quite stable, I have use it quite a lot with great success. No crash. I have fixed some issue mostly in the GUI part.

For redmine issue #10173

### CODE REVIEW

**Comment by MRIG** [comment]
first impression: meaningful tool
comment: when using line tool, the arrows on the line do look somewhat strange,like dots or smudge
the image must be increased to see that these are arrows that indicate the direction
another comment: the balloon explanation appears at the top of the image which is okay but
there you need to get used to it

**Comment by TurboGit** [comment]
Indeed most of the time you need to zoom, that's a bit like for the mask. It is sometime difficult to setup properly without zooming in the right area.

For the information this is the way "hint" messages are displayed in darktable.

And yes blending can give strange result has you are moving part of the image depending on the channel selected.

Thanks for the comments!

**Comment by TurboGit** [comment]
New version pushed without all commits to fix the iop. This should make the reviewer job easier.

**Comment by TurboGit** [comment]
Just rebased to clear conflics (with VNG OpenCL support) on master.

**Comment by TurboGit** [comment]
New version with all modifications (hope I did not forget one :) discussed on IRC. That is:
- use float and not double
- no more serialization code, just use a static array as in spots.c
- remove some part of code not used
- remove one-liner cairo routines
- remove javadoc style comments

Added:
- allow max 100 nodes, stop when limit reached
- display number of nodes/warps

This new implementation is 15k smaller than the original one :)

**Comment by LebedevRI** [comment]
@TurboGit how do you build darktable?
```
[ 70%] Building C object src/iop/CMakeFiles/liquify.dir/introspection_liquify.c.o
In file included from /home/lebedevri/darktable/build/src/iop/introspection_liquify.c:15:0:
/home/lebedevri/darktable/src/iop/liquify.c: In function ‘alloc_move_to’:
/home/lebedevri/darktable/src/iop/liquify.c:2334:10: error: return from incompatible pointer type [-Werror=incompatible-pointer-types]
   return m;
          ^
/home/lebedevri/darktable/src/iop/liquify.c: In function ‘alloc_line_to’:
/home/lebedevri/darktable/src/iop/liquify.c:2348:10: error: return from incompatible pointer type [-Werror=incompatible-pointer-types]
   return l;
          ^
/home/lebedevri/darktable/src/iop/liquify.c: In function ‘alloc_curve_to’:
/home/lebedevri/darktable/src/iop/liquify.c:2362:10: error: return from incompatible pointer type [-Werror=incompatible-pointer-types]
   return c;
          ^
cc1: all warnings being treated as errors
src/iop/CMakeFiles/liquify.dir/build.make:67: recipe for target 'src/iop/CMakeFiles/liquify.dir/introspection_liquify.c.o' failed
make[2]: *** [src/iop/CMakeFiles/liquify.dir/introspection_liquify.c.o] Error 1
CMakeFiles/Makefile2:4770: recipe for target 'src/iop/CMakeFiles/liquify.dir/all' failed
make[1]: *** [src/iop/CMakeFiles/liquify.dir/all] Error 2
Makefile:149: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by TurboGit** [comment]
Fixed.

**Comment by hanatos** [comment]
okay, since code comments in liquify.c don't seem to work, i'll try manually.

i have a comment about the comment in the code :) :

https://github.com/darktable-org/darktable/blob/liquify/src/iop/liquify.c#L407

all of this blurb about coordinate systems seems to be utterly wrong and makes me doubt the iop works correctly. i also observed some weird things when trying to zoom in on an image with applied transformations. these seem to switch themselves off and then be placed in a wrong place after zooming out again..?

first
* RAW: These are sensor coordinates. They go from x=0, y=0 to x=<sensor
 * width>, y=<sensor height>.
an iop has no idea about the raw dimensions, you'd have to pull it out from the mipmap cache. is that really what this thing tries to do? if so, it's broken. it should only care about the ROI and the corresponding full input buffer (buf_in). the raw has some black borders which are cropped away early in the pipe, for instance.

i'm not sure the reasoning with the distortions is very convincing. does this really work exactly? i mean, does it also convert strengths, falloffs, directions etc correctly? somehow i can't imagine this would work in the general case. if it doesn't work precisely the user will have to be aware of pipe ordering anyways to get accurate results. so maybe just ignore this?


https://github.com/darktable-org/darktable/blob/liquify/src/iop/liquify.c#L537
#define CAIRO_SCALE (1.0 / MAX (piece->pipe->backbuf_width, piece->pipe->backbuf_height))

is this ^ a hardcoded zoom-to fit scale? is that the reason why it does interesting things when i zoom in? btw it stays broken when i zoom back out.

**Comment by TurboGit** [comment]
Hum... This is a bug I have introduced in my reworked version as zoom in and out was working in the initial implementation. I've justed tested again with the original code and all is fine.

And yes this does work (actually the first implementation) as I have been using it :)

I need to review the recent changes...

**Comment by LebedevRI** [comment]
IMHO the first commit should be The original vanilla commit from @MarcelloPerathoner(?).
Because *this* code in pr is some variation of that original code.
Will be much easier to think about issues like those @hanatos pointed out.

**Comment by TurboGit** [comment]
Yes, you've already said that to me and I agree. I'll do that for my next push here.

**Comment by TurboGit** [comment]
Just pushed a new version with initial commit from Marcello and then my modifications. These version is fixed and works as expected.

File: tools/iop_dependencies.py [hunk 1]
```diff
@@ -204,7 +205,10 @@ def add_edges(gr):
   gr.add_edge(('lens', 'spots'))
   gr.add_edge(('borders', 'spots'))
   gr.add_edge(('clipping', 'spots'))
-  
+
+  # liquify immediately after spot removal
```
Comment (LebedevRI) [inline tools/iop_dependencies.py]: before-spots

**Comment by LebedevRI** [comment]
Does this module have any assumptions about it's working color space, does it need to work in camera rgb? lab?
Also, is it safe to assume that lens correction will be applied in direct direction (i.e. correct), not in reverse (i.e. distort)

Based on answers to these questions i see 3 possible places for liquify place in pipe:
1. before `spot removal`, like now
2. before `lens correction`
3. after `lens correction`, before input color profile
4. after `crop and rotate`?

Personally, 3. seems like the best place to me.
Or, at least comment in `tools/iop_dependencies.py` needs fixing :)

**Comment by hanatos** [comment]
On Mon, Jan 18, 2016 at 2:31 AM, Roman Lebedev <notifications@github.com>
wrote:

> Does this module have any assumptions about it's working color space, does
> it need to work in camera rgb? lab?
>
if anything, camera rgb (linear scene-referred light, we'll be taking
averages for pixel filtering). not sure it matters much.


> Also, is it safe to assume that lens correction will be applied in direct
> direction (i.e. correct), not in reverse (i.e. distort)
>
where is that assumption? it will evaluate the transforms of controls to
pixels via the respective distortion callbacks in lens.c, right? using
whatever parameters you set there?


> Based on answers to these questions i see 3 possible places for liquify
> place in pipe:
> 1. before spot removal, like now
> 2. before lens correction
> 3. after lens correction, before input color profile
>
i think 3 is the latest it could go (before input colour profile, linear
light).

and as disucssed in irc, i don't see a particular requirement. maybe if it
comes late in the workflow it should be later in the pipe (speed reasons),
but i can't judge liquify workflows.

**Comment by LebedevRI** [comment]
>where is that assumption? it will evaluate the transforms of controls to
>pixels via the respective distortion callbacks in lens.c, right? using
>whatever parameters you set there?

No, i meant "in most cases users will apply lens correction in direct direction"

>and as disucssed in irc, i don't see a particular requirement. maybe if it
>comes late in the workflow it should be later in the pipe (speed reasons),
>but i can't judge liquify workflows.

Agreed.

File: tools/iop_dependencies.py (hunk 1 above)
Comment (TurboGit) [inline tools/iop_dependencies.py]: comment fixed.

**Comment by TurboGit** [comment]
Hanatos and Roman are ok for merging after review. Tobias, we'd like your review before merging. If ok, let me know.

**Comment by LebedevRI** [comment]
(i did not say "i'm ok", i did say "i'm not against, after a brief look", there is a difference)

**Comment by TurboGit** [comment]
Ok, sorry Roman. That's what I meant. We just need another reviewer.

**Comment by TurboGit** [comment]
If there is no more objection I'll merge this at the end of the week-end.

**Comment by TurboGit** [comment]
Again some more testing an a minor issue fixed.

**Comment by LebedevRI** [comment]
Does not build with clang-3.8:
```
[ 70%] Building C object src/iop/CMakeFiles/liquify.dir/introspection_liquify.c.o
In file included from /home/lebedevri/darktable/build/src/iop/introspection_liquify.c:15:
/home/lebedevri/darktable/src/iop/liquify.c:1290:7: error: variable 'dev_roi_out' is used uninitialized whenever 'if' condition is true [-Werror,-Wsometimes-uninitialized]
  if (dev_roi_in == NULL) goto error;
      ^~~~~~~~~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1331:7: note: uninitialized use occurs here
  if (dev_roi_out   ) dt_opencl_release_mem_object (dev_roi_out);
      ^~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1290:3: note: remove the 'if' if its condition is always false
  if (dev_roi_in == NULL) goto error;
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1292:3: note: variable 'dev_roi_out' is declared here
  cl_mem_t dev_roi_out = dt_opencl_copy_host_to_device_constant (
  ^
/home/lebedevri/darktable/src/iop/liquify.c:1294:7: error: variable 'dev_map' is used uninitialized whenever 'if' condition is true [-Werror,-Wsometimes-uninitialized]
  if (dev_roi_out == NULL) goto error;
      ^~~~~~~~~~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1330:7: note: uninitialized use occurs here
  if (dev_map       ) dt_opencl_release_mem_object (dev_map);
      ^~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1294:3: note: remove the 'if' if its condition is always false
  if (dev_roi_out == NULL) goto error;
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1290:7: error: variable 'dev_map' is used uninitialized whenever 'if' condition is true [-Werror,-Wsometimes-uninitialized]
  if (dev_roi_in == NULL) goto error;
      ^~~~~~~~~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1330:7: note: uninitialized use occurs here
  if (dev_map       ) dt_opencl_release_mem_object (dev_map);
      ^~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1290:3: note: remove the 'if' if its condition is always false
  if (dev_roi_in == NULL) goto error;
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1296:3: note: variable 'dev_map' is declared here
  cl_mem_t dev_map = dt_opencl_copy_host_to_device_constant (
  ^
/home/lebedevri/darktable/src/iop/liquify.c:1298:7: error: variable 'dev_map_extent' is used uninitialized whenever 'if' condition is true [-Werror,-Wsometimes-uninitialized]
  if (dev_map == NULL) goto error;
      ^~~~~~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1329:7: note: uninitialized use occurs here
  if (dev_map_extent) dt_opencl_release_mem_object (dev_map_extent);
      ^~~~~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1298:3: note: remove the 'if' if its condition is always false
  if (dev_map == NULL) goto error;
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1294:7: error: variable 'dev_map_extent' is used uninitialized whenever 'if' condition is true [-Werror,-Wsometimes-uninitialized]
  if (dev_roi_out == NULL) goto error;
      ^~~~~~~~~~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1329:7: note: uninitialized use occurs here
  if (dev_map_extent) dt_opencl_release_mem_object (dev_map_extent);
      ^~~~~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1294:3: note: remove the 'if' if its condition is always false
  if (dev_roi_out == NULL) goto error;
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1290:7: error: variable 'dev_map_extent' is used uninitialized whenever 'if' condition is true [-Werror,-Wsometimes-uninitialized]
  if (dev_roi_in == NULL) goto error;
      ^~~~~~~~~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1329:7: note: uninitialized use occurs here
  if (dev_map_extent) dt_opencl_release_mem_object (dev_map_extent);
      ^~~~~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1290:3: note: remove the 'if' if its condition is always false
  if (dev_roi_in == NULL) goto error;
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1300:3: note: variable 'dev_map_extent' is declared here
  cl_mem_t dev_map_extent = dt_opencl_copy_host_to_device_constant (
  ^
/home/lebedevri/darktable/src/iop/liquify.c:1302:7: error: variable 'dev_kdesc' is used uninitialized whenever 'if' condition is true [-Werror,-Wsometimes-uninitialized]
  if (dev_map_extent == NULL) goto error;
      ^~~~~~~~~~~~~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1328:7: note: uninitialized use occurs here
  if (dev_kdesc     ) dt_opencl_release_mem_object (dev_kdesc);
      ^~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1302:3: note: remove the 'if' if its condition is always false
  if (dev_map_extent == NULL) goto error;
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1298:7: error: variable 'dev_kdesc' is used uninitialized whenever 'if' condition is true [-Werror,-Wsometimes-uninitialized]
  if (dev_map == NULL) goto error;
      ^~~~~~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1328:7: note: uninitialized use occurs here
  if (dev_kdesc     ) dt_opencl_release_mem_object (dev_kdesc);
      ^~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1298:3: note: remove the 'if' if its condition is always false
  if (dev_map == NULL) goto error;
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1294:7: error: variable 'dev_kdesc' is used uninitialized whenever 'if' condition is true [-Werror,-Wsometimes-uninitialized]
  if (dev_roi_out == NULL) goto error;
      ^~~~~~~~~~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1328:7: note: uninitialized use occurs here
  if (dev_kdesc     ) dt_opencl_release_mem_object (dev_kdesc);
      ^~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1294:3: note: remove the 'if' if its condition is always false
  if (dev_roi_out == NULL) goto error;
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1290:7: error: variable 'dev_kdesc' is used uninitialized whenever 'if' condition is true [-Werror,-Wsometimes-uninitialized]
  if (dev_roi_in == NULL) goto error;
      ^~~~~~~~~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1328:7: note: uninitialized use occurs here
  if (dev_kdesc     ) dt_opencl_release_mem_object (dev_kdesc);
      ^~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1290:3: note: remove the 'if' if its condition is always false
  if (dev_roi_in == NULL) goto error;
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1304:3: note: variable 'dev_kdesc' is declared here
  cl_mem_t dev_kdesc = dt_opencl_copy_host_to_device_constant (
  ^
/home/lebedevri/darktable/src/iop/liquify.c:1306:7: error: variable 'dev_kernel' is used uninitialized whenever 'if' condition is true [-Werror,-Wsometimes-uninitialized]
  if (dev_kdesc == NULL) goto error;
      ^~~~~~~~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1327:7: note: uninitialized use occurs here
  if (dev_kernel    ) dt_opencl_release_mem_object (dev_kernel);
      ^~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1306:3: note: remove the 'if' if its condition is always false
  if (dev_kdesc == NULL) goto error;
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1302:7: error: variable 'dev_kernel' is used uninitialized whenever 'if' condition is true [-Werror,-Wsometimes-uninitialized]
  if (dev_map_extent == NULL) goto error;
      ^~~~~~~~~~~~~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1327:7: note: uninitialized use occurs here
  if (dev_kernel    ) dt_opencl_release_mem_object (dev_kernel);
      ^~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1302:3: note: remove the 'if' if its condition is always false
  if (dev_map_extent == NULL) goto error;
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1298:7: error: variable 'dev_kernel' is used uninitialized whenever 'if' condition is true [-Werror,-Wsometimes-uninitialized]
  if (dev_map == NULL) goto error;
      ^~~~~~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1327:7: note: uninitialized use occurs here
  if (dev_kernel    ) dt_opencl_release_mem_object (dev_kernel);
      ^~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1298:3: note: remove the 'if' if its condition is always false
  if (dev_map == NULL) goto error;
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1294:7: error: variable 'dev_kernel' is used uninitialized whenever 'if' condition is true [-Werror,-Wsometimes-uninitialized]
  if (dev_roi_out == NULL) goto error;
      ^~~~~~~~~~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1327:7: note: uninitialized use occurs here
  if (dev_kernel    ) dt_opencl_release_mem_object (dev_kernel);
      ^~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1294:3: note: remove the 'if' if its condition is always false
  if (dev_roi_out == NULL) goto error;
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1290:7: error: variable 'dev_kernel' is used uninitialized whenever 'if' condition is true [-Werror,-Wsometimes-uninitialized]
  if (dev_roi_in == NULL) goto error;
      ^~~~~~~~~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1327:7: note: uninitialized use occurs here
  if (dev_kernel    ) dt_opencl_release_mem_object (dev_kernel);
      ^~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1290:3: note: remove the 'if' if its condition is always false
  if (dev_roi_in == NULL) goto error;
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/lebedevri/darktable/src/iop/liquify.c:1308:3: note: variable 'dev_kernel' is declared here
  cl_mem_t dev_kernel = dt_opencl_copy_host_to_device_constant (
  ^
15 errors generated.
src/iop/CMakeFiles/liquify.dir/build.make:67: recipe for target 'src/iop/CMakeFiles/liquify.dir/introspection_liquify.c.o' failed
make[2]: *** [src/iop/CMakeFiles/liquify.dir/introspection_liquify.c.o] Error 1
CMakeFiles/Makefile2:4770: recipe for target 'src/iop/CMakeFiles/liquify.dir/all' failed
make[1]: *** [src/iop/CMakeFiles/liquify.dir/all] Error 2
Makefile:149: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by TurboGit** [comment]
Thanks Roman. I'll try to fix all those. Since I'm not building with clang I cannot guarantee that it will build on clang when merging. I hope a good will with a clang compiler will help :)

**Comment by LebedevRI** [comment]
And the best question: what about `float complex`?
No other code does that.
Why?

**Comment by TurboGit** [comment]
clang issue should be fixed. easy, all in the same code.

**Comment by TurboGit** [comment]
complex is C99 standard and very handy for the computation in liquify. this has been discussed in IRC. the code will get messy without it.

**Comment by TurboGit** [comment]
BTW, what is the procedure to build using clang?

**Comment by LebedevRI** [comment]
Hmm, why does this reimplement interpolation.c, especially for opencl codepath.
@upegelow have you looked at this?

**Comment by TurboGit** [comment]
Liquify does use interpolation.c. It just redefine two small routines that are static in interpolation.c (bicubic and lanczos).

**Comment by TurboGit** [comment]
For the other part in apply_global_distortion_map_cl() it follows the same principle as in lenc.c and clipping.c.

**Comment by LebedevRI** [comment]
Hm?
Lens.c has a kernel per interpolation function, and does not re-implement interpolation functions.
Liquify.c does re-implement interpolation functions, and apply_global_distortion_map_cl() has only one kernel.

To me that seems like 2 exactly the opposite principles

**Comment by TurboGit** [comment]
Ok, not sure I follow :) I've asked Ulrich to have a look and will wait for his feedback before merging.

**Comment by TurboGit** [comment]
I've discussed this with Ulrich. The current implementation is fine. Apart from an unused routine that Ulrich spotted in liquify.cl all is ok. So I've fixed this and will now merge this PR!

**Comment by LebedevRI** [comment]
@TurboGit 
I feel like i should point out that `tools/iop_dependencies.py` places liquify into completely different place of pipe from where it is placed now, manually:
```
$ ../tools/iop_dependencies.py 
1000 gamma
984 dither
968 watermark
952 borders
936 finalscale
920 overexposed
904 clahe
888 splittoning
873 vignette
857 velvia
841 soften
825 channelmixer
809 colorout
793 colorcontrast
777 grain
761 highpass
746 lowpass
730 sharpen
714 colorcorrection
698 relight
682 levels
666 tonecurve
650 zonesystem
634 colisa
619 monochrome
603 lowlight
587 colorzones
571 bilat
555 atrous
539 shadhi
523 globaltonemap
507 nlmeans
492 bloom
476 colormapping
460 colortransfer
444 colorize
428 clipping
412 colorbalance
396 equalizer
380 defringe
365 colorreconstruction
349 colorin
333 profile_gamma
317 bilateral
301 basecurve
285 graduatednd
269 flip
253 liquify
238 lens
222 scalepixels
206 rotatepixels
190 spots
174 exposure
158 tonemap
142 denoiseprofile
126 demosaic
111 rawdenoise
95 hotpixels
79 cacorrect
63 highlights
47 temperature
31 invert
15 rawprepare
0 rawspeed
```

**Comment by TurboGit** [comment]
Never run this... Will fix.

**Comment by TurboGit** [comment]
@LebedevRI should be fixed. Thanks.

**Pull request merged**